File: c/sum-of-multiples/sum_of_multiples.c
#include "sum_of_multiples.h"

bool is_factor_of(unsigned int factor, unsigned int number);
bool has_factor_in(unsigned int number, const unsigned int *factors,
                   size_t number_of_factors);

unsigned int sum(const unsigned int *factors, const size_t number_of_factors,
                 const unsigned int limit) {
  unsigned int factors_sum = 0;
  for (size_t candidate = 1; candidate < limit; candidate++) {
    if (has_factor_in(candidate, factors, number_of_factors)) {
      factors_sum += candidate;
    }
  }
  return factors_sum;
}

bool is_factor_of(unsigned int factor, unsigned int number) {
  if (factor != 0) {
    return number % factor == 0;
  }
  if (number == 0) {
    return true;
  }
  return false;
}

bool has_factor_in(unsigned int number, const unsigned int *factors,
                   size_t number_of_factors) {
  for (size_t i = 0; i < number_of_factors; i++) {
    if (is_factor_of(factors[i], number)) {
      return true;
    }
  }
  return false;
}

File: java/sum-of-multiples/src/main/java/SumOfMultiples.java
import java.util.Arrays;
import java.util.Set;
import java.util.stream.Collectors;
import java.util.stream.IntStream;

class SumOfMultiples {

    private Integer sum;
    private Set<Integer> divisors;
    private final int limit;

    SumOfMultiples(int limit, int[] set) {
        this.limit = limit;
        this.divisors = Arrays.stream(set)
                .boxed()
                .collect(Collectors.toSet());

        this.sum = null;
    }

    int getSum() {
        if (this.sum != null) {
            return this.sum;
        }

        this.sum = IntStream.range(1, this.limit)
                .filter(this::hasDivisorsInSet)
                .sum();
        this.divisors = null;

        return this.sum;
    }

    private boolean hasDivisorsInSet(int number) {
        if (number == 0) {
            return true;
        }

        return this.divisors
                .stream()
                .anyMatch(d -> d != 0 && number % d == 0);
    }

}
